### DISCUSSION TITLE
SkiaSlider: Any way to disable parent page scrolling while interacting?

### DISCUSSION THREAD

**Discussion by NeighborhoodNerd** [taublast/DrawnUi#132] [Q&A] [answered]
Basically title. I'm struggling trying to figure out how to disable the parent ScrollView's scrolling behavior while interacting with the slider. When the user moves their finger up or down while holding onto the slider's thumb, the entire page scrolls up and down. In iOS, I was able to achieve this by binding to the IsEnabled property of the ScrollView but apparently that doesn't work in Android. 

I even found [this](https://www.sharpnado.com/net-maui-disable-scrolling-on-android/l) article that claims it can be done, but I get InvalidCast errors so this isn't an option for me. 

So, without adding a handler, is there a way to do this within SkiaSlider?

**Comment by taublast** [comment]
Hi are we talking about the context of 
A. MAUI ScrollView - StackLayout - Canvas - custom SkiaSlider
OR
B. Canvas - SkiaScroll -SkiaLayout - custom SkiaSlider
?

**Reply by NeighborhoodNerd** [reply]
Hi sorry this is a ScrollView > StackLayout > Canvas > Custom Slider

**Reply by taublast** [reply]
Please set `Gestures="Lock"` to your canvas and let me know if this works for you.

**Reply by NeighborhoodNerd** [reply]
I'll do that and let you know. Thanks! Btw, I'm loving this library. I was debating on switching to Avalonia but with the canvas view, it makes it easy to slowly transition my app over to drawn elements.

**Reply by taublast** [reply]
> I'll do that and let you know. Thanks! Btw, I'm loving this library. I was debating on switching to Avalonia but with the canvas view, it makes it easy to slowly transition my app over to drawn elements.

Thanks, yeah the main goal was to have a freedom of drawing anything instead of composing with native views + tons of renderers/handlers while still staying inside Xamarin/MAUI as at the time the alternative was like "go flutter". Dart???? Nah. :)

**Reply by NeighborhoodNerd** [reply]
> > I'll do that and let you know. Thanks! Btw, I'm loving this library. I was debating on switching to Avalonia but with the canvas view, it makes it easy to slowly transition my app over to drawn elements.
> 
> 
> 
> Thanks, yeah the main goal was to have a freedom of drawing anything instead of composing with native views + tons of renderers/handlers while still staying inside Xamarin/MAUI as at the time the alternative was like "go flutter". Dart???? Nah. :)

Yeah I totally agree. Us C# devs need more options and this totally fits the bill. Avalonia Maui hybrid only supports mobile platforms, and vanilla Avalonia has almost no support for Shell and Navigation views so being able to host this within Maui is the best of both worlds.

**Comment by taublast** [comment]
While testing i see that the built-it "ignore wrong direction" behavior is not showing well for this case. 
I mean in case of Maui scrollview we pan the slider lets say "vertically" it ignores those gestures, while when in "gestures locked" context this is not good at all. I will add a property to ignore or NOT the wrong direction, same as we have for skiascroll, so the solution for such case would be to add gestures = lock to canvas and ignorewrongdirection=false to slider. The new nuget will be out today, along we some other new stuff.

**Reply by NeighborhoodNerd** [reply]
Ah ok. I did try the IgnoreWrongDirection property. Thanks a lot for the quick support. I'll look forward to that nuget then :)

**Reply by taublast** [reply]
> Ah ok. I did try the IgnoreWrongDirection property. Thanks a lot for the quick support. I'll look forward to that nuget then :)

It seems to not work well in case of of lock with MAUI so I even thought it was not implemented here at all. Well, solving this anyway right now, until it's comfortable to pan drawn slider inside native scrollview. I had it solved already looong ago when I was just including canvases here and there with drawn sliders, buttons etc inside Xamarin, so we can be sure it's possible as it was already done just with a different implementation. :)

**Comment by taublast** [comment]
Latest nuget https://github.com/taublast/DrawnUi.Maui/discussions/134 should give canvas priority over scrollview with `Lock` option. 

If you have small canvases with controls adding Gestures="Lock" should solve the problem. So basically we have 2 working modes: normal and lock. With normal you could pan scrollview having finger on the canvas, still passing gestures to canvas. With locked you could not pan scrollview when finger is on the canvas. 

I had a third mode "share" that I had implemented in xamarin, the basic idea is that drawn control was reporting back to gesture recognizer whether it has consumed gesture or not. So in case of vertical scroll it could report NOT and scrollview could then scroll vertically. I cannot implement this here at this time, because gestures are processed in async manner, canvas is not reporting back anything to native gesture recognizers.

I see maybe a way of still implementing this but not in a very near future.

**Reply by NeighborhoodNerd** [reply]
This is working beautifully. Thank's so much for getting this fixed as quickly as you did!

**Reply by NeighborhoodNerd** [reply]
I may have jumped the gun responding. iOS will not navigate to a page that has a Canvas with anything other than Gestures="Disabled" now. Android is working fine.

**Reply by taublast** [reply]
"iOS will not navigate to a page" , - what is it precisely, Maui Shell, or?.. 
Would it be possible to ask You for some repro code?

**Reply by NeighborhoodNerd** [reply]
When I set Gestures="Enabled", I get a crash with this Stack Trace:

System.InvalidOperationException: No service for type 'Microsoft.Maui.Controls.Hosting.EffectsFactory' has been registered.
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[EffectsFactory](IServiceProvider provider)
   at Microsoft.Maui.Controls.Element.get_EffectsFactory()
   at Microsoft.Maui.Controls.Element.Microsoft.Maui.Controls.IEffectControlProvider.RegisterEffect(Effect effect)
   at Microsoft.Maui.Controls.Element.AttachEffect(Effect effect)
   at Microsoft.Maui.Controls.Element.set_EffectControlProvider(IEffectControlProvider value)
   at Microsoft.Maui.Controls.Element.OnHandlerChangedCore()

**Reply by NeighborhoodNerd** [reply]
```<?xml version="1.0" encoding="utf-8"?>

<ContentPage xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             xmlns:controls="clr-namespace:TestApp.Controls"
             x:Class="TestApp.Pages.UITestPage"
             Style="{StaticResource GradientBackgroundPageStyle}"
             xmlns:pages="clr-namespace:TestApp.Pages"
             x:DataType="pages:UITestPageViewModel"
             xmlns:views="http://schemas.appomobi.com/drawnUi/2023/draw"
             Loaded="UITestPage_OnLoaded"
             Shell.NavBarIsVisible="False"
             Shell.PresentationMode="Animated">
    <Grid RowDefinitions="60,*">
        <controls:PageTitleBar Title="UI Test Page"/>
        <Grid Grid.Row="1">
            <ScrollView>
                <VerticalStackLayout>
                    <views:Canvas HorizontalOptions="Fill" Margin="8,0,8,0" HeightRequest="100" BackgroundColor="Green" x:Name="mainCanvas" Gestures="Enabled">
                        <controls:ColorSlider
                            Tag="Primary"
                            HeightRequest="50">
                        </controls:ColorSlider>
                    </views:Canvas>
                </VerticalStackLayout>  
            </ScrollView>
        </Grid>

    </Grid>

</ContentPage>`

**Reply by NeighborhoodNerd** [reply]
And for whatever reason, I can't get anything to render within Canvas using this test page..it works in my other pages in my app but this TestPage will not render anything within the canvas view, I've tried SkiaButton, SkiaImage, and this custom SkiaSlider

**Reply by taublast** [reply]
Very interesting... What i need to find out here is how you navigate to this page, how is it created comparing to other pages. What makes it different from others?

What happens if you do not enable gestures in xaml here, and in the OnAppearing set on UI thread mainCanvas.Gestures = TouchMode.Lock ?

**Reply by NeighborhoodNerd** [reply]
```    
[RelayCommand]
public async Task OpenSupportPage()
{
    #if ENV_Test
    await Shell.Current.GoToAsync(nameof(UITestPage));
    #else
    Shell.Current.CurrentItem = Shell.Current.Items.FirstOrDefault(x => x.Title == AppResource.SupportNavTitle);
    #endif
}`

When I try to set Gestures to Lock from the code-behind, it throws the same exception at that line, so no difference in behavior.

**Reply by taublast** [reply]
Normally a page is created with all services ready inside its Handler.MauiContext. 
We have here "System.InvalidOperationException: No service for type 'Microsoft.Maui.Controls.Hosting.EffectsFactory' has been registered." 
Means the code for setting the property value Gestures="whatever" is executed BEFORE the MauiContext was fully populated as for other working cases set. 
Why shell provides a handler here without a context missing our effects?.. Maui shell code.
How to solve? The key here is to understand the difference in instantiation by shell between working pages and this broken case, i'm still missing this info too..

**Reply by NeighborhoodNerd** [reply]
To clarify, it is working on Android, even when using my UITestPage I gave the example for. With setting Lock in the code-behind or xaml. This is iOS only that is not working.

**Reply by taublast** [reply]
But other pages are fine on iOS? How do you instantiate them? Same as the buggy one or?..

**Reply by NeighborhoodNerd** [reply]
No, if I set Gestures to anything other than Disabled, the page just refuses to navigate. So, they're definitely not fine. But for whatever reason the exception is being swallowed in other cases. I've had this happen when debugging my app if XAML is missing a style or something to that effect, the compiler doesn't complain but the page just never navigates. These other pages are being instantiated the same way. They're not injected or anything, just navigating via Shell.

**Comment by NeighborhoodNerd** [comment]
I’m using MAUI Shell, yeah. I’ll throw together an example ContentPage.

**Comment by taublast** [comment]
I'll try to reproduce, these days, meanwhile, I know this is much to ask but if you could create a small repro would be just awesome. If not, no worries ill deal with what I have.

**Reply by NeighborhoodNerd** [reply]
https://github.com/nathenxbrewer/DrawnUICanvas

I can't even run this repo on iOS. Getting the same EffectsFactory exception. Not even navigating to the page. This is a new > MAUI project with DrawnUi 1.2.9.4 installed and a Canvas wrapping a SkiaButton

**Comment by taublast** [comment]
what is the .NET version and what is the iOS version?

**Reply by NeighborhoodNerd** [reply]
.NET 8.0.402, iOS 18.0

**Comment by taublast** [comment]
Also can you delete your bin obj? Could be old version assemblies bugging with the new nuget

**Reply by NeighborhoodNerd** [reply]
I gave the reproduction repo. That's a fresh project and doesn't work.

**Comment by taublast** [comment]
I have cloned your repo, then have run it on emulator iOS 18.1, it clicks np on the red canvas

**Comment by taublast** [comment]
also might be maui workload issue. I do not have 9 on purpose:

Installed Workload Id      Manifest Version      Installation Source
--------------------------------------------------------------------
maui-maccatalyst           8.0.83/8.0.100        SDK 8.0.200        
maui-ios                   8.0.83/8.0.100        SDK 8.0.200        
maui-android               8.0.83/8.0.100        SDK 8.0.200

**Reply by NeighborhoodNerd** [reply]
This worked for the Simulator with the sample repo I gave, but having nothing but issues on my retail app. Not sure why or how they are different, trying to figure that out now. 

registrar.h:1411:43: warning: 'retain (or strong)' attribute on property 'window' does not match the property inherited from 'UIApplicationDelegate' [-Wproperty-attribute-mismatch]
 1411 |         @property (nonatomic, assign) UIWindow * window;

**Reply by NeighborhoodNerd** [reply]
I'm not sure what I did but I can't run any of my MAUI apps anymore..

Workload version: 8.0.400

Installed Workload Id      Installation Source
----------------------------------------------
maui-ios                   SDK 8.0.400        
maui-android               SDK 8.0.400        
ios                        SDK 8.0.400        
maui                       SDK 8.0.400        
android                    SDK 8.0.400        
wasm-tools-net7            SDK 8.0.400        

Use `dotnet workload search` to find additional workloads to install.

nathenbrewer@Nathens-Macbook-Pro-50 ~ % dotnet --list-sdks
6.0.418 [/usr/local/share/dotnet/sdk]
6.0.422 [/usr/local/share/dotnet/sdk]
6.0.423 [/usr/local/share/dotnet/sdk]
6.0.425 [/usr/local/share/dotnet/sdk]
7.0.316 [/usr/local/share/dotnet/sdk]
7.0.317 [/usr/local/share/dotnet/sdk]
8.0.404 [/usr/local/share/dotnet/sdk]


I'm on MacOS 15.1

**Reply by NeighborhoodNerd** [reply]
Yeah, my entire dev environment is screwed at this point. I've re-installed Xcode, downgraded to 8.0.200, re-installed all the maui workloads, and now Visual Studio builds successfully but all of my apps just insta-crash on both my iPhones. I'll have to let you know if I can get my environment working if I can get your package to work.

**Reply by NeighborhoodNerd** [reply]
I cannot use 8.0.100. When I do, it says I need Xcode 16, then I change my Xcode path to 16, and it gives me the register.h errors. None of my MAUI apps work with this SDK. If I update to 8.0.404, all of my apps work fine but your package still crashes the app and throws the EffectsFactory error. Not sure what to try next.

**Reply by NeighborhoodNerd** [reply]
In fact, I can't use any version newer than 1.2.3.8 for this package. All versions above it throw that same exception.

**Comment by taublast** [comment]
Just in case this is due to maui version, could use a smaller one: https://github.com/dotnet/maui/issues/12998

**Comment by taublast** [comment]
Oh so sorry to hear that. If your apps are running fine on simulators and they crash only on real devices maybe we should also check from the profiles and certificates side.

**Reply by taublast** [reply]
Also maybe the difference with the real devices is they get some internet data from Wi-Fi that is not working for them and they crash up on it? I would just check devices console logs, for the crash message. Or maybe I misunderstood and this is still that registrar error message? Do your apps without my lib run fine on devices?

**Reply by NeighborhoodNerd** [reply]
They're crashing on both physical and simulators. It's not your library that I can tell. It's definitely something with the MAUI version vs Xcode version vs MacOS version. I'll get it straightened out and let you know how the Canvas works after that.

**Comment by taublast** [comment]
Okay so what is the status? Do I understand correctly that:
1 yor apps are working fine with some SDK combination, but
2 adding a new drawnui version upper than 1.2.3.8 will fail adding gestures effect on iOS
Correct?

**Reply by NeighborhoodNerd** [reply]
Correct, all of my apps work when I'm on 8.0.404, but because I'm on Sequoia 15.1, I'm having issues running any app with SDK 8.0.200. If I add anything newer than 1.2.3.8 to my project, it throws the exception if Gestures is anything other than Disabled.

**Reply by taublast** [reply]
Interesting im on sequoya with  xcode 16 and 8.0.83/8.0.100 SDK 8.0.200 and M1. Im compiling and running with VS Code, rider fails me. Ill try run something with the latest nuget tomorrow, a new nuget will come out tonight, not sure it would fix anything but ill try repro with it.

**Reply by NeighborhoodNerd** [reply]
I couldn’t run any Maui app with that config. I’m on an M3 arm processor. Either way, I lost almost a full day of work so I’ll need to  just try this again sometime I guess I’ll just exclude this functionality from the Android version of my app for now. I have it working  on iOS by just disabling the scrollview.

**Comment by taublast** [comment]
I see the main difference is our setup is VS Code, if I understand correctly:
You: vscode 16.0.
Me: vscode 16.1, can use simulators in VS Code for Mac only. Now I'm scared to downgrade to 16.0, am I right?

**Reply by NeighborhoodNerd** [reply]
I tried Visual Studio for Mac, Rider, and VS Code with SDK 8.0.200 and couldn't even compile any of my MAUI apps. I currently have XCode 15.4 and 16 Beta installed. I have to have 15.4 installed even though technically it's not supported on Sequoia, otherwise I can't build iOS.

**Comment by taublast** [comment]
Have some news. I wanted to debug a Xamarin app, so that led to making installs on Mac Sequoya:

1. XCODE 15.4
* With Xcodes app installed Xcode 15.4
* then made in runnable on this OS, terminal: defaults write /Applications/Xcode-15.4.0.app/Contents/Info.plist CFBundleVersion -string 23051
* then activated it and installed platform 17.5 still in Xcodes app

2. for MAUI
* installed older MAUI workload to work with Xcode 15.4 with sdk 17.5:
sudo dotnet workload install maui-ios --from-rollback-file https://github.com/dotnet/maui/releases/download/8.0.82/8.0.82.json
* installed older Rider version 2023.3.6 from https://www.jetbrains.com/rider/download/other.html hoping for it to work with an old Xcode and MAUI
* Now Rider can see simulators, can't see physical device for MAUI.

4. for XAMARIN 
* Xamarin.iOS 16.4.0.18 https://github.com/xamarin/xamarin-macios/blob/main/DOWNLOADS.md
* Found Visual Studio for Mac on web
* Now VS for Mac can see simulators and physical device for xamarin, debugs fine

Now regarding our case:

When launching MAUI app on simulator first crash is due to bin/obj not deleted by hands.
After that it crashes with default	21:22:37.456696+0300	Sandbox	
Unhandled Exception:
System.InvalidOperationException: No service for type 'Microsoft.Maui.Controls.Hosting.EffectsFactory' has been registered.
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[EffectsFactory](IServiceProvider provider)
   at Microsoft.Maui.Controls.Element.get_EffectsFactory()
   at Microsoft.Maui.Controls.Element.Microsoft.Maui.Controls.IEffectControlProvider.RegisterEffect(Effect effect)
   at Microsoft.Maui.Controls.Element.AttachEffect(Effect effect)
   at Microsoft.Maui.Controls.Element.set_EffectControlProvider(IEffectControlProvider value)
   at Microsoft.Maui.Controls.Element.OnHandlerChangedCore()
   at Microsoft.Maui.Controls.VisualElement.OnHandlerChangedCore()
   at Microsoft.Maui.Controls.View.OnHandlerChangedCore()
   at Microsoft.Maui.Controls.Element.SetHandler(IElementHandle<…>

Can investigate that now

**Comment by taublast** [comment]
The silly explanation to the bug. left is the new nuget crashing on older ios workloads. right is the old nuget.
![image](https://github.com/user-attachments/assets/ffd8d1cd-0fb4-4ba7-bd77-8c579844a4b0)

**Reply by taublast** [reply]
You could manually reference fixed gestures until dranwui nuget gets updated:

`dotnet add package AppoMobi.Maui.Gestures --version 1.9.2.5`

**Reply by taublast** [reply, marked as answer]
@nathenxbrewer the new drawnui stable(!) nuget should fix all issues building for lower ios versions. do not use preview pls.

**Reply by NeighborhoodNerd** [reply]
@taublast this is working great now! Thanks!
